File: guides/micronaut-openapi-base/java/src/main/java/example/micronaut/Guide.java
/*
 * Copyright 2017-2024 original authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package example.micronaut;

import io.micronaut.core.annotation.Nullable;
import io.micronaut.serde.annotation.Serdeable;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import java.time.LocalDate;
import java.util.List;

@Serdeable // <1>
public record Guide(@NotBlank String title,
                    @NotBlank String intro,
                    @NotNull @Size(min = 1) List<@NotBlank String> authors, // <2>
                    @Nullable List<@NotBlank String> tags,
                    @NotNull @Size(min = 1) List<@NotBlank String> categories,
                    @Schema(format = "yyyy-MM-dd", example = "2018-05-23") @NotNull LocalDate publicationDate, // <3>
                    @NotBlank String slug,
                    @NotBlank String url,
                    @NotNull @Size(min = 1) List<@NotNull @Valid Option> options) {
}

File: guides/micronaut-openapi-base/python/src/example/micronaut/guide.py
from dataclasses import dataclass
from typing import Annotated

import java
from jakarta.validation import Valid
from jakarta.validation.constraints import NotBlank, NotNull, Size
from java.time import LocalDate
from micronaut.serde.annotation import Serdeable

from .option import Option


Schema = java.type("io.swagger.v3.oas.annotations.media.Schema")


@Serdeable  # <1>
@dataclass
class Guide:
    title: Annotated[str, NotBlank]
    intro: Annotated[str, NotBlank]
    authors: Annotated[list[str], NotNull, Size(min=1)]  # <2>
    tags: list[str] | None
    categories: Annotated[list[str], NotNull, Size(min=1)]
    publicationDate: Annotated[
        LocalDate,
        NotNull,
        Schema(format="yyyy-MM-dd", example="2018-05-23"),
    ]  # <3>
    slug: Annotated[str, NotBlank]
    url: Annotated[str, NotBlank]
    options: Annotated[list[Option], NotNull, Size(min=1), Valid]
